charon/src/remove_read_discriminant.rs

```diff
@@ -102,53 +69,106 @@ impl<'a, 'tcx, 'ctx> Visitor<'a, 'tcx, 'ctx> {
+                        // be of any *signed* integer type (`isize`, `i8`, etc.).
+                        let discr_to_id: HashMap<ScalarValue, VariantId::Id> = variants
+                            .iter_indexed_values()
+                            .map(|(id, variant)| (variant.discriminant, id))
+                            .collect();
+                        let mut covered_discriminants: HashSet<ScalarValue> = HashSet::default();
+                        let targets = targets
+                            .into_iter()
+                            .map(|(v, e)| {
+                                (
+                                    v.into_iter()
+                                        .filter_map(|discr| {
+                                            covered_discriminants.insert(discr);
+                                            discr_to_id.get(&discr).or_else(|| {
+                                                register_error_or_panic!(
+                                                    self.ctx,
+                                                    st.meta.span.rust_span_data.span(),
+                                                    "Found incorrect discriminant {discr} for enum {adt_id}"
+                                                );
+                                                None
+                                            })
+                                        })
+                                        .copied()
+                                        .collect_vec(),
+                                    e,
+                                )
+                            })
+                            .collect_vec();
+                        // Filter the otherwise branch if it is not necessary.
+                        let covers_all = covered_discriminants.len() == discr_to_id.len();
+                        let otherwise = if covers_all { None } else { Some(otherwise) };
+
+                        let switch =
+                            RawStatement::Switch(Switch::Match(p.clone(), targets, otherwise));
+
+                        // Add the next statement if there is one
+                        st.content = if let Some(st3) = st3_opt {
+                            let meta = combine_meta(meta1, &meta2);
+                            let switch = Statement {
+                                meta,
+                                content: switch,
+                            };
+                            new_sequence(switch, st3).content
+                        } else {
+                            switch
+                        };
+                    }
+                    Err(st2) => {
+                        // The discriminant read is not followed by a `SwitchInt`. We replace `_x =
+                        // Discr(_y)` with `match _y { 0 => { _x = 0 }, 1 => { _x = 1; }, .. }`.
+                        let targets = variants
```

Comment:
Just add a comment which states that it happens when we do, for instance: `y as usize`.

Reply:
That's not the case. As far as I know only optimizing a `match` can make that happen. I'll note that

Reply:
Ok, then just add a comment which explains when this can happen :)